src/python/pants/core/goals/test.py

```diff
@@ -287,6 +287,9 @@ async def run_tests(
     )
 
     exit_code = PANTS_SUCCEEDED_EXIT_CODE
+
+    console.teardown_dynamic_ui()
```

Comment:
Rather than having rules call this explicitly, it feels like the first writes to stdio per `@goal_rule` could do this automatically?

Reply:
The only challenge is that we hold onto the same `Console` object for an entire session, which might end up running multiple `@goal_rules`: https://github.com/pantsbuild/pants/blob/master/src/python/pants/init/engine_initializer.py#L174

Reply:
> The only challenge is that we hold onto the same `Console` object for an entire session, which might end up running multiple `@goal_rules`: https://github.com/pantsbuild/pants/blob/master/src/python/pants/init/engine_initializer.py#L174

Hm, I think that implies that we also need a way for goal rules to bring the dynamic UI back up? I don't think it matters much that the Python method happens to be exposed on a `Console` object, what matters is the session, since that's where the rust dynamic UI code all lives.

Reply:
It doesn't if each new `@goal_rule` restarts it, which is what will currently happen, because each `@goal_rule` is a new run of the engine.